core/tests/performance_framework/perf_services.py

```diff
@@ -0,0 +1,227 @@
+
+from core.tests.performance_framework import perf_domain
+
+CHROME_DRIVER_PATH = os.path.join(
+    '..', 'node_modules', 'protractor', 'selenium', 'chromedriver_2.21')
+
+BROWSERMOB_PROXY_PATH = os.path.join(
+    '..', 'oppia_tools', 'browsermob-proxy-2.1.1', 'bin', 'browsermob-proxy')
+
+
+class SeleniumPerformanceDataFetcher(object):
+    """Fetches performance data for locally served Oppia pages using Selenium
+    and Browsermob-proxy.
+
+    Selenium is used to programitically interact with a browser.
+    Browsermob-proxy is used to capture HTTP Archive (referred to as HAR) data.
+
+    The HTTP Archive format is a JSON-formatted archive file format used for
+    logging a web browser's interaction with a site. It contains detailed
+    performance data, including information about page loading and displaying
+    and per-resource statistics. Each entry contains the URL requested and
+    request and response headers. For additional details, please see:
+    https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/HAR/Overview.html
+
+    The HTTP Archive data is retrieved using the proxy server by recording the
+    communication between the client and server.
+    """
+
+    # Chrome is the default browser that selenium will use for extracting
+    # performance stats.
+    DEFAULT_BROWSER_SOURCE = 'chrome'
+    SUPPORTED_BROWSER_SOURCES = ['chrome', 'firefox']
+
+    # Duration (in seconds) to wait for the complete page to load, otherwise
+    # XHR requests made post initial page load will not be recorded.
+    WAIT_DURATION_SECS = 3
+
+    def __init__(self, browser=DEFAULT_BROWSER_SOURCE):
+        if browser in self.SUPPORTED_BROWSER_SOURCES:
+            self.browser = browser
+        else:
+            error_msg = 'Unsupported browser specified: %s' % browser
+            raise ValueError(error_msg)
+
+        self.server = self.driver = self.proxy = None
+
+    def get_page_metrics_for_url(self, page_url):
+        """Returns a PageSessionMetrics domain object for a given page URL.
+        """
+        with Xvfb() as _:
+            self._setup_proxy()
+            self._setup_driver(use_proxy=True)
+
+            self.proxy.new_har(page_url, options={'captureHeaders': True})
+
+            self.driver.get(page_url)
+
+            self._wait_until_page_load_is_finished()
+
+        har_dict = self.proxy.har
+
+        self._stop_proxy()
+        self._stop_driver()
+
+        return perf_domain.PageSessionMetrics(
+            page_session_stats=har_dict, page_session_timings=None)
+
+    def get_page_metrics_from_cached_session(self, page_url):
+        """Returns a PageSessionMetrics domain object for a given page URL
+        while simulating a cached session i.e, a return user.
+
+        Note: This method is stand-alone and does not require
+        `get_page_session_metrics_for_url` to be called before it.
+        """
+        with Xvfb() as _:
+            self._setup_proxy()
+            self._setup_driver(use_proxy=True)
```

Comment:
Feel free to ignore, thinking out loud:
if we change _setup_driver to return driver instead of assigning to self.driver, we will have one less state to keep in this class.